Mathlib/Data/Pi/Algebra.lean

```diff
@@ -0,0 +1,398 @@
+
+variable (x y : ∀ i, f i) (i : I)
+
+namespace Pi
+
+/-! `1`, `0`, `+`, `*`, `+ᵥ`, `•`, `^`, `-`, `⁻¹`, and `/` are defined pointwise. -/
+
+@[to_additive]
+instance instOne [∀ i, One <| f i] : One (∀ i : I, f i) :=
+  ⟨fun _ => 1⟩
+
+#align pi.has_one Pi.instOne
+
+@[simp, to_additive]
+theorem one_apply [∀ i, One <| f i] : (1 : ∀ i, f i) i = 1 :=
+  rfl
+
+@[to_additive]
+theorem one_def [∀ i, One <| f i] : (1 : ∀ i, f i) = fun _ => 1 :=
+  rfl
+
+@[simp, to_additive]
+theorem const_one [One β] : const α (1 : β) = 1 :=
+  rfl
+
+@[simp, to_additive]
+theorem one_comp [One γ] (x : α → β) : (1 : β → γ) ∘ x = 1 :=
+  rfl
+
+@[simp, to_additive]
+theorem comp_one [One β] (x : β → γ) : x ∘ (1 : α → β) = const α (x 1) :=
+  rfl
+
+@[to_additive]
+instance instMul [∀ i, Mul <| f i] : Mul (∀ i : I, f i) :=
+  ⟨fun f g i => f i * g i⟩
+
+#align pi.has_mul Pi.instMul
+
+@[simp, to_additive]
+theorem mul_apply [∀ i, Mul <| f i] : (x * y) i = x i * y i :=
+  rfl
+
+@[to_additive]
+theorem mul_def [∀ i, Mul <| f i] : x * y = fun i => x i * y i :=
+  rfl
+
+@[simp, to_additive]
+theorem const_mul [Mul β] (a b : β) : const α a * const α b = const α (a * b) :=
+  rfl
+
+@[to_additive]
+theorem mul_comp [Mul γ] (x y : β → γ) (z : α → β) : (x * y) ∘ z = x ∘ z * y ∘ z :=
+  rfl
+
+@[to_additive]
+instance instHasSmul [∀ i, HasSmul α <| f i] : HasSmul α (∀ i : I, f i) :=
+  ⟨fun s x => fun i => s • x i⟩
+
+@[simp, to_additive]
+theorem smul_apply [∀ i, HasSmul α <| f i] (s : α) (x : ∀ i, f i) (i : I) : (s • x) i = s • x i :=
+  rfl
+
+@[to_additive]
+theorem smul_def [∀ i, HasSmul α <| f i] (s : α) (x : ∀ i, f i) : s • x = fun i => s • x i :=
+  rfl
+
+@[simp, to_additive]
+theorem smul_const [HasSmul α β] (a : α) (b : β) : a • const I b = const I (a • b) :=
+  rfl
+
+@[to_additive]
+theorem smul_comp [HasSmul α γ] (a : α) (x : β → γ) (y : I → β) : (a • x) ∘ y = a • x ∘ y :=
+  rfl
+
+@[to_additive Pi.instHasSmul]
+instance instPow [∀ i, Pow (f i) β] : Pow (∀ i, f i) β :=
+  ⟨fun x b i => x i ^ b⟩
+
+@[simp, to_additive Pi.smul_apply, to_additive_reorder 5]
+theorem pow_apply [∀ i, Pow (f i) β] (x : ∀ i, f i) (b : β) (i : I) : (x ^ b) i = x i ^ b :=
+  rfl
+
+@[to_additive Pi.smul_def, to_additive_reorder 5]
+theorem pow_def [∀ i, Pow (f i) β] (x : ∀ i, f i) (b : β) : x ^ b = fun i => x i ^ b :=
+  rfl
+
+-- `to_additive` generates bad output if we take `Pow α β`.
+@[simp, to_additive smul_const, to_additive_reorder 5]
+theorem const_pow [Pow β α] (b : β) (a : α) : const I b ^ a = const I (b ^ a) :=
+  rfl
+
+@[to_additive smul_comp, to_additive_reorder 6]
+theorem pow_comp [Pow γ α] (x : β → γ) (a : α) (y : I → β) : (x ^ a) ∘ y = x ∘ y ^ a :=
+  rfl
+
+-- -- Porting note : deprecated. not ported.
+-- @[simp]
+-- theorem bit0_apply [∀ i, Add <| f i] : (bit0 x) i = bit0 (x i) :=
+--   rfl
+--
+-- @[simp]
+-- theorem bit1_apply [∀ i, Add <| f i] [∀ i, One <| f i] : (bit1 x) i = bit1 (x i) :=
+--   rfl
```

Comment:
I think we've been porting those anyway with `set_option linter.deprecated false` to avoid the warnings

Reply:
done.